server/vulnerabilities/oval/analyzer.go

```diff
@@ -0,0 +1,175 @@
+			if len(toDelete) > 0 {
+				if err := ds.DeleteVulnerabilitiesByCPECVE(ctx, toDelete); err != nil {
+					return nil, err
+				}
+			}
+			if len(toInsert) > 0 {
+				if _, err = ds.InsertVulnerabilities(ctx, toInsert, fleet.OVAL); err != nil {
+					return nil, err
+				}
+				return toInsert, nil
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+// vulnsDelta compares what vulnerabilities already exists with what new vulnerabilities were found
+// and returns what to insert and what to delete.
+func vulnsDelta(
+	found []fleet.SoftwareVulnerability,
+	existing []fleet.SoftwareVulnerability,
+) ([]fleet.SoftwareVulnerability, []fleet.SoftwareVulnerability) {
+	toDelete := make([]fleet.SoftwareVulnerability, 0)
+	toInsert := make([]fleet.SoftwareVulnerability, 0)
+
+	existingSet := make(map[string]bool)
+	for _, e := range existing {
+		existingSet[e.CVE] = true
+	}
+
+	foundSet := make(map[string]bool)
+	for _, f := range found {
+		foundSet[f.CVE] = true
+	}
+
+	for _, e := range existing {
+		if _, ok := foundSet[e.CVE]; !ok {
+			toDelete = append(toDelete, e)
+		}
+	}
+
+	for _, f := range found {
+		if _, ok := existingSet[f.CVE]; !ok {
+			toInsert = append(toInsert, f)
+		}
+	}
+
+	return toInsert, toDelete
+}
+
+// loadDef returns the latest oval Definition for the given platform.
+func loadDef(platform Platform, vulnPath string) (oval_parsed.Result, error) {
+	latest, err := latestOvalDefFor(platform, vulnPath, time.Now())
+	if err != nil {
+		return nil, err
+	}
+	paylaod, err := ioutil.ReadFile(latest)
+	if err != nil {
+		return nil, err
+	}
+
+	if platform.IsUbuntu() {
+		result := oval_parsed.UbuntuResult{}
+		if err := json.Unmarshal(paylaod, &result); err != nil {
+			return nil, err
+		}
+		return result, nil
+	}
+
+	return nil, fmt.Errorf("don't know how to load OVAL file for '%s' platform", platform)
+}
+
+// latestOvalDefFor returns the contents of the OVAL definition for the given 'platform' in
+// 'vulnPath' for the given 'date'.
+// If not found, returns the most up to date OVAL definition for the given 'platform'
+func latestOvalDefFor(platform Platform, vulnPath string, date time.Time) (string, error) {
+	ext := "json"
+	fileName := platform.ToFilename(date, ext)
+	target := path.Join(vulnPath, fileName)
+
+	_, err := os.Stat(target)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			files, err := os.ReadDir(vulnPath)
+			if err != nil {
+				return "", err
+			}
+
+			prefix := strings.Split(fileName, "-")[0]
+			var latest os.FileInfo
+			for _, f := range files {
+				if strings.HasPrefix(f.Name(), prefix) && strings.HasSuffix(f.Name(), ext) {
+					info, err := f.Info()
+					if err != nil {
+						continue
+					}
+					if latest == nil || info.ModTime().After(latest.ModTime()) {
+						latest = info
+					}
+				}
+			}
+
+			if latest != nil {
+				return path.Join(vulnPath, latest.Name()), nil
+			}
+		}
+		return "", fmt.Errorf("file not found for platform '%s' in '%s'", platform, vulnPath)
```

Comment:
This `return` can hit if `if err != nil && !errors.Is(err, fs.ErrNotExist)`, so the error is not accurate.

Reply:
Alternative: https://github.com/fleetdm/fleet/pull/5797/files#r878566004